polkadot/node/core/runtime-api/src/lib.rs

```diff
@@ -327,6 +327,24 @@ where
 		}
 	}
 
+	// Checks that all set features in runtime are known by the current version of the node, if not
+	// it throws an warning for users to update their version.
+	async fn maybe_check_node_features(
+		should_check_node_features: Option<(Hash, SessionIndex, Arc<Client>, Metrics)>,
+	) {
+		if let Some((block_hash, session_index, client, metrics)) = should_check_node_features {
+			let (tx, _rx) = oneshot::channel();
+			let request = Request::NodeFeatures(session_index, tx);
+			let result = make_runtime_api_request(client, metrics, block_hash, request).await;
```

Comment:
should we first check the cache?

Reply:
yeah, but they're different caches, there's one LruMap for session info and another one for NodeFeatures. Since node features are session buffered it should be the same though, given that we configure the same cache capacity